src/error.rs

```diff
@@ -1,39 +1,124 @@
 #![allow(overflowing_literals)]
 
+use crate::runtime::*;
+use crate::*;
+
+/// The ErrorCode (a.k.a HRESULT) of an error
+#[repr(transparent)]
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub struct ErrorCode(pub i32);
+
+/// A WinRT related error
+pub struct Error {
+    code: ErrorCode,
+    info: IRestrictedErrorInfo,
+}
+
 /// An alias for `std::result::Result<T, winrt::Error>`
 #[must_use]
 pub type Result<T> = std::result::Result<T, Error>;
 
 impl<T> std::convert::From<Result<T>> for ErrorCode {
     fn from(result: Result<T>) -> Self {
         if let Err(error) = result {
-            // TODO: call SetErrorInfo
+            if !error.info.is_null() {
+                unsafe {
+                    SetRestrictedErrorInfo(error.info.as_raw() as _);
+                }
+            }
             return error.code();
         }
 
         ErrorCode(0)
     }
 }
 
-/// A WinRT related error
-#[derive(Debug)]
-pub struct Error {
-    code: ErrorCode,
-    // TODO: add `info: IErrorInfo`
+impl std::convert::From<ErrorCode> for Error {
+    fn from(code: ErrorCode) -> Self {
+        unsafe {
+            let mut info = IErrorInfo::default();
+            GetErrorInfo(0, info.set_abi() as _);
+            let restricted = info.query::<IRestrictedErrorInfo>();
+
+            if !restricted.is_null() {
+                let info2 = info.query::<ILanguageExceptionErrorInfo2>();
+
+                if !info2.is_null() {
+                    info2.capture_propagation_context();
+                }
+
+                return Self {
+                    code,
+                    info: restricted,
+                };
+            }
+
+            let mut message = String::new();
+
+            if !info.is_null() {
```

Comment:
All the null checking is very annoying especially since the compiler won't enforce it. More reason to have us be able to model WInRT types as not nullable.